docs/dev/kube-apiserver-health-check.md

```diff
@@ -0,0 +1,70 @@
+- The load balancer detects it and takes it out of service in time. No new request should be forwarded to the 
+  `kube-apiserver` instance when it has stopped listening.
+- Existing connections are not cut off hard, they are allowed to complete gracefully.
+
+## Load Balancer Health Check Probe
+`kube-apiserver` provides graceful termination support via the `/readyz` health check endpoint. When `/readyz`reports
+`ok` it indicates that the apiserver is ready to serve request(s).
+
+Now let's walk through the events (in chronological order) that unfold when a `kube-apiserver` instance restarts:
+* E1: `T+0s`: `kube-apiserver` receives a TERM signal.
+* E2: `T+0s`: `/readyz` starts reporting `failure` to signal to the load balancer that a shut down is in progress.
+  * The apiserver will continue to accept new request(s).
+  * The apiserver waits for certain amount of time (configurable by `shutdown-delay-duration`) before it stops accepting new request(s).
+* E3: `T+70s`: `kube-apiserver` (the http server) stops listening:
+  * `/healthz` turns red.
+  * Default TCP health check probe on port `6443` will fail.
+  * Any new request forwarded to it will fail, most likely with a `connection refused` error.
+  * Existing request(s) in-flight are not cut off but are allowed to complete gracefully, but there is a timeout of `60s`.
+* E4: `T+70s+60s`: Any existing request(s) that are still in flight are terminated.
+* E5: `T+70s+60s`: The apiserver process exits.
+
+An important note to consider is that today the time difference between `E3` and `E2` is `70s`. This is known as
+`shutdown-delay-duration` and is configurable.
+```
+$ kubectl -n openshift-kube-apiserver get cm config -o json | jq -r '.data."config.yaml"' |
+  jq '.apiServerArguments."shutdown-delay-duration"'
+[
+  "70s"
+]
+```
+In future we will reduce `shutdown-delay-duration` to `30s`. So for the sake of this document let's assume that `E3 - E2` is `30s`.
+
+Given the above, we can infer the following:
+* The load balancer should use `/readyz` endpoint for `kube-apiserver` health check probe. It must NOT use `/healthz` or
+default TCP port probe.
+* The time taken by the load balancer (let's say `t` seconds) to deem a `kube-apiserver` instance unhealthy and take it
+out of service should not bleed into `E3`. So `E2 + t < E3` must be true so that no new request is forwarded to the instance at `E3` or later. In the worst case it should
+take `30s` (since `E2` triggers) for the load balancer to take the `kube-apiserver` instance out of service.
+
+Below is the health check configuration used by `aws` currently.
+
+```
+protocol: HTTPS
+path: /readyz
+port: 6443
+unhealthy threshold: 2
+timeout: 10s
+interval: 10s
+```
+
+Let's plot the events as it happens when a load balance determines that a `kube-apiserver` is unhealthy and 
+takes it out of service. This considers a worst case scenario:
+* P1: T+0s:  `/readyz` starts reporting red.
+* P2: T+10s: first probe initiated.
```

Comment:
be clear that this is an example. P2 could be right at T+0s, depending on the alignment of the probe request interval.

Reply:
made it clear that this is a worst case scenario to calculate `at most 30s`